File: scripts/ai-workflow/phases/test_implementation.py
"""Phase 5: テストコード実装フェーズ

Phase 3で作成されたテストシナリオとPhase 4で実装された実コードを基に、
テストコードのみを実装する。実コードの修正は行わない。
"""
import re
from pathlib import Path
from typing import Dict, Any, List
from .base_phase import BasePhase


class TestImplementationPhase(BasePhase):
    """テストコード実装フェーズ"""

    def __init__(self, *args, **kwargs):
        """初期化"""
        super().__init__(
            phase_name='test_implementation',
            *args,
            **kwargs
        )

    def execute(self) -> Dict[str, Any]:
        """
        テストコード実装フェーズを実行

        Returns:
            Dict[str, Any]: 実行結果
                - success: bool
                - output: str - test-implementation.mdのパス
                - error: Optional[str]
        """
        try:
            # Issue情報を取得
            issue_number = int(self.metadata.data['issue_number'])

            # 要件定義書、設計書、テストシナリオ、実装ログを読み込み
            requirements_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '01_requirements' / 'output' / 'requirements.md'
            design_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '02_design' / 'output' / 'design.md'
            test_scenario_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '03_test_scenario' / 'output' / 'test-scenario.md'
            implementation_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '04_implementation' / 'output' / 'implementation.md'

            # ファイル存在確認
            missing_files = []
            if not requirements_file.exists():
                missing_files.append(f'要件定義書: {requirements_file}')
            if not design_file.exists():
                missing_files.append(f'設計書: {design_file}')
            if not test_scenario_file.exists():
                missing_files.append(f'テストシナリオ: {test_scenario_file}')
            if not implementation_file.exists():
                missing_files.append(f'実装ログ: {implementation_file}')

            if missing_files:
                return {
                    'success': False,
                    'output': None,
                    'error': f'必要なファイルが見つかりません:\n' + '\n'.join(missing_files)
                }

            # テスト戦略を取得（Phase 2で決定済み）
            test_strategy = self.metadata.data['design_decisions'].get('test_strategy')
            test_code_strategy = self.metadata.data['design_decisions'].get('test_code_strategy')

            if not test_strategy or not test_code_strategy:
                return {
                    'success': False,
                    'output': None,
                    'error': 'テスト戦略が設計フェーズで決定されていません。Phase 2を先に実行してください。'
                }

            # Planning Phase成果物のパス取得
            planning_path_str = self._get_planning_document_path(issue_number)

            # 実行プロンプトを読み込み
            execute_prompt_template = self.load_prompt('execute')

            # working_dirからの相対パスを使用
            rel_path_requirements = requirements_file.relative_to(self.claude.working_dir)
            rel_path_design = design_file.relative_to(self.claude.working_dir)
            rel_path_test_scenario = test_scenario_file.relative_to(self.claude.working_dir)
            rel_path_implementation = implementation_file.relative_to(self.claude.working_dir)

            # プロンプトに情報を埋め込み
            execute_prompt = execute_prompt_template.replace(
                '{planning_document_path}',
                planning_path_str
            ).replace(
                '{requirements_document_path}',
                f'@{rel_path_requirements}'
            ).replace(
                '{design_document_path}',
                f'@{rel_path_design}'
            ).replace(
                '{test_scenario_document_path}',
                f'@{rel_path_test_scenario}'
            ).replace(
                '{implementation_document_path}',
                f'@{rel_path_implementation}'
            ).replace(
                '{test_strategy}',
                test_strategy
            ).replace(
                '{test_code_strategy}',
                test_code_strategy
            ).replace(
                '{issue_number}',
                str(issue_number)
            )

            # Claude Agent SDKでタスクを実行
            # テスト実装フェーズは時間がかかる可能性があるため、max_turnsを多めに
            messages = self.execute_with_claude(
                prompt=execute_prompt,
                max_turns=50,
                log_prefix='execute'
            )

            # test-implementation.mdのパスを取得
            output_file = self.output_dir / 'test-implementation.md'

            if not output_file.exists():
                return {
                    'success': False,
                    'output': None,
                    'error': f'test-implementation.mdが生成されませんでした: {output_file}'
                }

            # GitHub Issueに成果物を投稿
            try:
                output_content = output_file.read_text(encoding='utf-8')
                self.post_output(
                    output_content=output_content,
                    title="テストコード実装ログ"
                )
            except Exception as e:
                print(f"[WARNING] 成果物のGitHub投稿に失敗しました: {e}")

            return {
                'success': True,
                'output': str(output_file),
                'error': None
            }

        except Exception as e:
            # ステータス更新: 失敗
            self.metadata.update_phase_status('test_implementation', 'failed')

            return {
                'success': False,
                'output': None,
                'error': str(e)
            }

    def review(self) -> Dict[str, Any]:
        """
        テストコード実装をレビュー

        Returns:
            Dict[str, Any]: レビュー結果
                - result: str - PASS/PASS_WITH_SUGGESTIONS/FAIL
                - feedback: str
                - suggestions: List[str]
        """
        try:
            # test-implementation.mdを読み込み
            test_implementation_file = self.output_dir / 'test-implementation.md'

            if not test_implementation_file.exists():
                return {
                    'result': 'FAIL',
                    'feedback': 'test-implementation.mdが存在しません。',
                    'suggestions': ['execute()を実行してtest-implementation.mdを生成してください。']
                }

            # 設計書、テストシナリオ、実装ログのパス
            issue_number = int(self.metadata.data['issue_number'])
            design_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '02_design' / 'output' / 'design.md'
            test_scenario_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '03_test_scenario' / 'output' / 'test-scenario.md'
            implementation_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '04_implementation' / 'output' / 'implementation.md'

            # テスト戦略を取得
            test_strategy = self.metadata.data['design_decisions'].get('test_strategy', 'UNKNOWN')
            test_code_strategy = self.metadata.data['design_decisions'].get('test_code_strategy', 'UNKNOWN')

            # レビュープロンプトを読み込み
            review_prompt_template = self.load_prompt('review')

            # working_dirからの相対パスを使用
            rel_path_test_implementation = test_implementation_file.relative_to(self.claude.working_dir)
            rel_path_design = design_file.relative_to(self.claude.working_dir)
            rel_path_test_scenario = test_scenario_file.relative_to(self.claude.working_dir)
            rel_path_implementation = implementation_file.relative_to(self.claude.working_dir)

            # プロンプトに情報を埋め込み
            review_prompt = review_prompt_template.replace(
                '{test_implementation_document_path}',
                f'@{rel_path_test_implementation}'
            ).replace(
                '{design_document_path}',
                f'@{rel_path_design}'
            ).replace(
                '{test_scenario_document_path}',
                f'@{rel_path_test_scenario}'
            ).replace(
                '{implementation_document_path}',
                f'@{rel_path_implementation}'
            ).replace(
                '{test_strategy}',
                test_strategy
            ).replace(
                '{test_code_strategy}',
                test_code_strategy
            )

            # Claude Agent SDKでレビューを実行
            messages = self.execute_with_claude(
                prompt=review_prompt,
                max_turns=30,
                log_prefix='review'
            )

            # レビュー結果をパース
            review_result = self._parse_review_result(messages)

            # レビュー結果をファイルに保存
            review_file = self.review_dir / 'result.md'
            review_file.write_text(review_result['feedback'], encoding='utf-8')
            print(f"[INFO] レビュー結果を保存: {review_file}")

            return review_result

        except Exception as e:
            return {
                'result': 'FAIL',
                'feedback': f'レビュー中にエラーが発生しました: {str(e)}',
                'suggestions': []
            }

    def revise(self, review_feedback: str) -> Dict[str, Any]:
        """
        レビュー結果を元にテストコードを修正

        Args:
            review_feedback: レビュー結果のフィードバック

        Returns:
            Dict[str, Any]: 修正結果
                - success: bool
                - output: str - test-implementation.mdのパス
                - error: Optional[str]
        """
        try:
            # 元のテスト実装ログを読み込み
            test_implementation_file = self.output_dir / 'test-implementation.md'

            if not test_implementation_file.exists():
                return {
                    'success': False,
                    'output': None,
                    'error': 'test-implementation.mdが存在しません。'
                }

            # 設計書、テストシナリオ、実装ログのパス
            issue_number = int(self.metadata.data['issue_number'])
            design_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '02_design' / 'output' / 'design.md'
            test_scenario_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '03_test_scenario' / 'output' / 'test-scenario.md'
            implementation_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '04_implementation' / 'output' / 'implementation.md'

            # テスト戦略を取得
            test_strategy = self.metadata.data['design_decisions'].get('test_strategy', 'UNKNOWN')
            test_code_strategy = self.metadata.data['design_decisions'].get('test_code_strategy', 'UNKNOWN')

            # 修正プロンプトを読み込み
            revise_prompt_template = self.load_prompt('revise')

            # working_dirからの相対パスを使用
            rel_path_test_implementation = test_implementation_file.relative_to(self.claude.working_dir)
            rel_path_design = design_file.relative_to(self.claude.working_dir)
            rel_path_test_scenario = test_scenario_file.relative_to(self.claude.working_dir)
            rel_path_implementation = implementation_file.relative_to(self.claude.working_dir)

            # プロンプトに情報を埋め込み
            revise_prompt = revise_prompt_template.replace(
                '{test_implementation_document_path}',
                f'@{rel_path_test_implementation}'
            ).replace(
                '{review_feedback}',
                review_feedback
            ).replace(
                '{design_document_path}',
                f'@{rel_path_design}'
            ).replace(
                '{test_scenario_document_path}',
                f'@{rel_path_test_scenario}'
            ).replace(
                '{implementation_document_path}',
                f'@{rel_path_implementation}'
            ).replace(
                '{test_strategy}',
                test_strategy
            ).replace(
                '{test_code_strategy}',
                test_code_strategy
            ).replace(
                '{issue_number}',
                str(issue_number)
            )

            # Claude Agent SDKでタスクを実行
            messages = self.execute_with_claude(
                prompt=revise_prompt,
                max_turns=50,
                log_prefix='revise'
            )

            # test-implementation.mdのパスを取得
            output_file = self.output_dir / 'test-implementation.md'

            if not output_file.exists():
                return {
                    'success': False,
                    'output': None,
                    'error': '修正されたtest-implementation.mdが生成されませんでした。'
                }

            return {
                'success': True,
                'output': str(output_file),
                'error': None
            }

        except Exception as e:
            return {
                'success': False,
                'output': None,
                'error': str(e)
            }

File: scripts/ai-workflow-v2/src/phases/test-implementation.ts
import fs from 'fs-extra';
import path from 'node:path';
import { BasePhase, type PhaseInitializationParams } from './base-phase.js';
import { PhaseExecutionResult } from '../types.js';

export class TestImplementationPhase extends BasePhase {
  constructor(params: PhaseInitializationParams) {
    super({ ...params, phaseName: 'test_implementation' });
  }

  protected async execute(): Promise<PhaseExecutionResult> {
    const issueNumber = parseInt(this.metadata.data.issue_number, 10);
    const planningReference = this.getPlanningDocumentReference(issueNumber);

    // オプショナルコンテキストを構築（Issue #398）
    const requirementsContext = this.buildOptionalContext(
      'requirements',
      'requirements.md',
      '要件定義書は利用できません。Planning情報とIssue情報から要件を推測してください。',
      issueNumber,
    );

    const designContext = this.buildOptionalContext(
      'design',
      'design.md',
      '設計書は利用できません。Issue情報とPlanning情報に基づいて適切な設計判断を行ってください。',
      issueNumber,
    );

    const scenarioContext = this.buildOptionalContext(
      'test_scenario',
      'test-scenario.md',
      'テストシナリオは利用できません。実装時に適切なテスト考慮を行ってください。',
      issueNumber,
    );

    const implementationContext = this.buildOptionalContext(
      'implementation',
      'implementation.md',
      '実装ログは利用できません。設計書とテストシナリオに基づいて実装してください。',
      issueNumber,
    );

    // test_strategy と test_code_strategy もオプショナル（Issue #405）
    const testStrategy = this.metadata.data.design_decisions.test_strategy ??
      'テスト戦略は設定されていません。設計書とテストシナリオから適切なテスト戦略を決定してください。';
    const testCodeStrategy = this.metadata.data.design_decisions.test_code_strategy ??
      'テストコード方針は設定されていません。プロジェクトの規約とテスト戦略から適切なテストコード方針を決定してください。';

    const executePrompt = this.loadPrompt('execute')
      .replace('{planning_document_path}', planningReference)
      .replace('{requirements_context}', requirementsContext)
      .replace('{design_context}', designContext)
      .replace('{test_scenario_context}', scenarioContext)
      .replace('{implementation_context}', implementationContext)
      .replace('{test_strategy}', testStrategy)
      .replace('{test_code_strategy}', testCodeStrategy)
      .replace('{issue_number}', String(issueNumber));

    await this.executeWithAgent(executePrompt, { maxTurns: 40 });

    const testImplementationFile = path.join(this.outputDir, 'test-implementation.md');
    if (!fs.existsSync(testImplementationFile)) {
      return {
        success: false,
        error: `test-implementation.md が見つかりません: ${testImplementationFile}`,
      };
    }

    // Phase outputはPRに含まれるため、Issue投稿は不要（Review resultのみ投稿）
    // try {
    //   const content = fs.readFileSync(testImplementationFile, 'utf-8');
    //   await this.postOutput(content, 'テストコード実装方針');
    // } catch (error) {
    //   const message = (error as Error).message ?? String(error);
    //   console.warn(`[WARNING] GitHub へのテストコード実装出力に失敗しました: ${message}`);
    // }

    return {
      success: true,
      output: testImplementationFile,
    };
  }

  protected async review(): Promise<PhaseExecutionResult> {
    const issueNumber = parseInt(this.metadata.data.issue_number, 10);
    const testImplementationFile = path.join(this.outputDir, 'test-implementation.md');

    if (!fs.existsSync(testImplementationFile)) {
      return {
        success: false,
        error: 'test-implementation.md が存在しません。execute() を先に実行してください。',
      };
    }

    const planningReference = this.getPlanningDocumentReference(issueNumber);
    const designFile = this.getPhaseOutputFile('design', 'design.md', issueNumber);
    const scenarioFile = this.getPhaseOutputFile('test_scenario', 'test-scenario.md', issueNumber);
    const implementationFile = this.getPhaseOutputFile('implementation', 'implementation.md', issueNumber);

    const testImplementationReference = this.getAgentFileReference(testImplementationFile);
    if (!testImplementationReference) {
      return {
        success: false,
        error: 'Agent が test-implementation.md を参照できません。',
      };
    }

    // design, scenario, implementation はオプショナル（Issue #405）
    let designReference: string;
    if (designFile) {
      const ref = this.getAgentFileReference(designFile);
      designReference = ref ?? '設計ドキュメントは利用できません。テストコード実装内容から設計を推測してレビューしてください。';
    } else {
      designReference = '設計ドキュメントは利用できません。テストコード実装内容から設計を推測してレビューしてください。';
    }

    let scenarioReference: string;
    if (scenarioFile) {
      const ref = this.getAgentFileReference(scenarioFile);
      scenarioReference = ref ?? 'テストシナリオは利用できません。テストコード実装内容から適切なテスト観点でレビューしてください。';
    } else {
      scenarioReference = 'テストシナリオは利用できません。テストコード実装内容から適切なテスト観点でレビューしてください。';
    }

    let implementationReference: string;
    if (implementationFile) {
      const ref = this.getAgentFileReference(implementationFile);
      implementationReference = ref ?? '実装ログは利用できません。テストコード実装内容から実装を推測してレビューしてください。';
    } else {
      implementationReference = '実装ログは利用できません。テストコード実装内容から実装を推測してレビューしてください。';
    }

    const testStrategy = this.metadata.data.design_decisions.test_strategy ??
      'テスト戦略は設定されていません。テストコード実装内容から適切なテスト観点でレビューしてください。';
    const testCodeStrategy = this.metadata.data.design_decisions.test_code_strategy ??
      'テストコード方針は設定されていません。プロジェクトの規約に基づいてレビューしてください。';

    const reviewPrompt = this.loadPrompt('review')
      .replace('{planning_document_path}', planningReference)
      .replace('{test_implementation_document_path}', testImplementationReference)
      .replace('{design_document_path}', designReference)
      .replace('{test_scenario_document_path}', scenarioReference)
      .replace('{implementation_document_path}', implementationReference)
      .replace('{test_strategy}', testStrategy)
      .replace('{test_code_strategy}', testCodeStrategy);

    const messages = await this.executeWithAgent(reviewPrompt, { maxTurns: 30, logDir: this.reviewDir });
    const reviewResult = await this.contentParser.parseReviewResult(messages);

    const reviewFile = path.join(this.reviewDir, 'result.md');
    fs.writeFileSync(reviewFile, reviewResult.feedback, 'utf-8');

    await this.github.postReviewResult(
      issueNumber,
      this.phaseName,
      reviewResult.result,
      reviewResult.feedback,
      reviewResult.suggestions,
    );

    return {
      success: reviewResult.result !== 'FAIL',
      output: reviewResult.result,
      error: reviewResult.result === 'FAIL' ? reviewResult.feedback : undefined,
    };
  }

  public async revise(reviewFeedback: string): Promise<PhaseExecutionResult> {
    const issueNumber = parseInt(this.metadata.data.issue_number, 10);
    const testImplementationFile = path.join(this.outputDir, 'test-implementation.md');

    if (!fs.existsSync(testImplementationFile)) {
      return {
        success: false,
        error: 'test-implementation.md が存在しません。execute() を先に実行してください。',
      };
    }

    const designFile = this.getPhaseOutputFile('design', 'design.md', issueNumber);
    const scenarioFile = this.getPhaseOutputFile('test_scenario', 'test-scenario.md', issueNumber);
    const implementationFile = this.getPhaseOutputFile('implementation', 'implementation.md', issueNumber);

    const testImplementationReference = this.getAgentFileReference(testImplementationFile);
    if (!testImplementationReference) {
      return {
        success: false,
        error: 'Agent が test-implementation.md を参照できません。',
      };
    }

    // design, scenario, implementation はオプショナル（Issue #405）
    let designReference: string;
    if (designFile) {
      const ref = this.getAgentFileReference(designFile);
      designReference = ref ?? '設計ドキュメントは利用できません。テストコード実装内容から設計を推測してください。';
    } else {
      designReference = '設計ドキュメントは利用できません。テストコード実装内容から設計を推測してください。';
    }

    let scenarioReference: string;
    if (scenarioFile) {
      const ref = this.getAgentFileReference(scenarioFile);
      scenarioReference = ref ?? 'テストシナリオは利用できません。テストコード実装内容から適切なテスト観点で修正してください。';
    } else {
      scenarioReference = 'テストシナリオは利用できません。テストコード実装内容から適切なテスト観点で修正してください。';
    }

    let implementationReference: string;
    if (implementationFile) {
      const ref = this.getAgentFileReference(implementationFile);
      implementationReference = ref ?? '実装ログは利用できません。テストコード実装内容から実装を推測してください。';
    } else {
      implementationReference = '実装ログは利用できません。テストコード実装内容から実装を推測してください。';
    }

    const testStrategy = this.metadata.data.design_decisions.test_strategy ??
      'テスト戦略は設定されていません。テストコード実装内容から適切なテスト戦略を決定してください。';
    const testCodeStrategy = this.metadata.data.design_decisions.test_code_strategy ??
      'テストコード方針は設定されていません。プロジェクトの規約とテスト戦略から適切なテストコード方針を決定してください。';

    const revisePrompt = this.loadPrompt('revise')
      .replace('{test_implementation_document_path}', testImplementationReference)
      .replace('{design_document_path}', designReference)
      .replace('{test_scenario_document_path}', scenarioReference)
      .replace('{implementation_document_path}', implementationReference)
      .replace('{test_strategy}', testStrategy)
      .replace('{test_code_strategy}', testCodeStrategy)
      .replace('{review_feedback}', reviewFeedback)
      .replace('{issue_number}', String(issueNumber));

    await this.executeWithAgent(revisePrompt, { maxTurns: 50, logDir: this.reviseDir });

    if (!fs.existsSync(testImplementationFile)) {
      return {
        success: false,
        error: '改訂後の test-implementation.md を確認できませんでした。',
      };
    }

    return {
      success: true,
      output: testImplementationFile,
    };
  }
}
